packages/opentelemetry-resources/src/Resource.ts

```diff
@@ -71,85 +58,82 @@ export class Resource implements IResource {
+      const [k,v] = this._rawAttributes[i];
+      try {
+        this._rawAttributes[i] = [k, await v];
+      } catch (err) {
         diag.debug("a resource's async attributes promise rejected: %s", err);
-        this.asyncAttributesPending = false;
-        return {};
+        this._rawAttributes[i] = [k, undefined];
       }
-    );
+    }
+
+    this._asyncAttributesPending = false;
   }
 
-  get attributes(): Attributes {
+  public get attributes(): Attributes {
     if (this.asyncAttributesPending) {
       diag.error(
         'Accessing resource attributes before async attributes settled'
       );
     }
 
-    return this._attributes ?? {};
-  }
+    if (this._memoizedAttributes) {
+      return this._memoizedAttributes;
+    }
 
-  /**
-   * Returns a promise that will never be rejected. Resolves when all async attributes have finished being added to
-   * this Resource's attributes. This is useful in exporters to block until resource detection
-   * has finished.
-   */
-  async waitForAsyncAttributes?(): Promise<void> {
-    if (this.asyncAttributesPending) {
-      await this._asyncAttributesPromise;
+    const attrs: Attributes = {};
+    for (const [k, v] of this._rawAttributes) {
+      if (isPromiseLike(v)) {
+        diag.debug(`Unsettled resource attribute ${k} skipped`);
+        continue;
+      }
+      if (v != null) {
+        attrs[k] ??= v;
+      }
     }
-  }
 
-  /**
-   * Returns a new, merged {@link Resource} by merging the current Resource
-   * with the other Resource. In case of a collision, other Resource takes
-   * precedence.
-   *
-   * @param other the Resource that will be merged with this.
-   * @returns the newly merged Resource.
-   */
-  merge(other: IResource | null): IResource {
-    if (!other) return this;
-
-    // Attributes from other resource overwrite attributes from this resource.
-    const mergedSyncAttributes = {
-      ...this._syncAttributes,
-      //Support for old resource implementation where _syncAttributes is not defined
-      ...((other as Resource)._syncAttributes ?? other.attributes),
-    };
-
-    if (
-      !this._asyncAttributesPromise &&
-      !(other as Resource)._asyncAttributesPromise
-    ) {
-      return new Resource(mergedSyncAttributes);
+    // only memoize output if all attributes are settled
+    if (!this._asyncAttributesPending) {
+      this._memoizedAttributes = attrs;
     }
 
-    const mergedAttributesPromise = Promise.all([
-      this._asyncAttributesPromise,
-      (other as Resource)._asyncAttributesPromise,
-    ]).then(([thisAsyncAttributes, otherAsyncAttributes]) => {
-      return {
-        ...this._syncAttributes,
-        ...thisAsyncAttributes,
-        //Support for old resource implementation where _syncAttributes is not defined
-        ...((other as Resource)._syncAttributes ?? other.attributes),
-        ...otherAsyncAttributes,
-      };
-    });
+    return attrs;
+  }
+
+  public merge(resource: Resource | null) {
```

Comment:
Hmm, I think here we may run into a #5283-like situation, we won't be able to merge resources of different package versions.

Reply:
Yes that's true. If we want to make this interface-only it would fix that, but then we would have to expose some getter for attributes that includes unresolved ones. Might be fine though.

Reply:
This will be fixed by https://github.com/open-telemetry/opentelemetry-js/issues/5358